### PR TITLE
Add the relocation_model to the cfg

### PR DESCRIPTION
**Pull request by lu-zero**
This way is possible to write inline assembly code aware of it.

### CODE REVIEW

**Review by Urgau — COMMENTED** [review]
(disclaimer: not a rust-lang reviewer)

Looks good, although this will probably need to be gated under a feature gate [[1]](https://github.com/rust-lang/rust/blob/8164cdb9ee460eddcb8888b9b6f72836a956d110/compiler/rustc_feature/src/builtin_attrs.rs#L24-L38) [[2]](https://github.com/rust-lang/rust/blob/8164cdb9ee460eddcb8888b9b6f72836a956d110/compiler/rustc_feature/src/active.rs#L127) [[3]](https://rustc-dev-guide.rust-lang.org/feature-gates.html#adding-a-feature-gate) with a [UI test](https://rustc-dev-guide.rust-lang.org/tests/ui.html).

**Comment by bjorn3** [comment]
I just thought of a possible issue. There is no guarantee in which crate a function is codegened, but `#[cfg]` always applies to the current crate. As such if the crate from which the `asm!()` originated and the one in which it was codegened have a different `-Crelocation-model` value you would get inline asm compiled with the wrong relocation model. This is even more of a problem with LTO where everything is effectively codegened in the executable, cdylib or staticlib crate.

**Comment by lu-zero** [comment]
The information is mainly useful to correctly consume symbols made available via `sym`, not sure if somebody has other use for it.

**Comment by bjorn3** [comment]
If you have

```rust
//- lib.rs compileflags:-Crelocation-model=static -Clto=yes --edition 2021
static FOO: u8 = 42;
pub unsafe fn foo() {
    #[cfg(relocation_model = "static")]
    core::arch::asm!("mov dword ptr [{}], 43", sym FOO);

    #[cfg(relocation_model = "pic")]
    core::arch::asm!("mov dword ptr [rip + {}], 43", sym FOO);
}

//- bar.rs compileflags:-Crelocation-model=pic -Clto=yes --extern lib=liblib.rlib --edition 2021
fn main() {
    unsafe { lib::foo(); }
}
```

Then `core::arch::asm!("mov dword ptr [{}], 43", sym FOO);` will be compiled with the PIC relocation model, while `core::arch::asm!("mov dword ptr [rip + {}], 43", sym FOO);` is necessary to get it working.

**Comment by bjorn3** [comment]
In most cases `-Crelocation-model` for all crates will match, but not always. And likely not for the standard library.

**Comment by lu-zero** [comment]
clang and gcc with any bit of inline asm with "m" constraint have to deal with it already, what do they do?

**Comment by bjorn3** [comment]
For LTO I think just accept that it happens and for inline and generic functions in C and C++ you need to include the header file, which will evaluate the `#ifdef` depending on the current compiler invocation and for regular functions it is guaranteed that any codegen will happen locally for non-LTO builds. For rustc all `#[cfg]` however are fixed at the point of producing the crate metadata that stands in for header files in C and C++.

**Comment by lu-zero** [comment]
the equivalent C code is

``` c
// lib.c - cc -c lib.c -fno-lto -flto -o lib.o
unsigned char FOO = 42;

void foo() {
#ifdef __PIC__
...
#else 
...
#endif
}
```
``` c
// main.c - cc -c main.c -flto -fPIC -o main.o
void foo();

int main() {
    foo();
    return 0;
}
```

``` 
# cc -flto main.o lib.o -o out
```
```
$ cc -fno-pic -c lib.c -o lib.o
$ cc -flto -c main.c -o main.o
$ cc -flto main.o lib.o -o out
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: lib.o: relocation R_X86_64_32S against symbol `FOO' can not be used when making a PIE object; recompile with -fPIE
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
```

The linker simply bails.

**Comment by Amanieu** [comment]
Generally you would never have a library that is "less PIC" than the binary using it. So this case should never happen in practice if you using a build system like Cargo.

**Comment by b-naber** [comment]
r? compiler

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1217,6 +1218,7 @@ fn default_configuration(sess: &Session) -> CrateConfig {
     ret.insert((sym::target_pointer_width, Some(Symbol::intern(&wordsz))));
     ret.insert((sym::target_env, Some(Symbol::intern(env))));
     ret.insert((sym::target_abi, Some(Symbol::intern(abi))));
+    ret.insert((sym::relocation_model, Some(relocation_model)));
```
Comment (bjorn3) [inline compiler/rustc_session/src/config.rs:1221]: This should be gated behind `sess.is_nightly_build()` to avoid immediately stabilizing it.

**Review by bjorn3 — CHANGES_REQUESTED** [review]
This will need a test.

File: compiler/rustc_session/src/config.rs (hunk 1 above)
Comment (lu-zero) [reply compiler/rustc_session/src/config.rs]: Done and added a test to make sure as well.

File: compiler/rustc_feature/src/active.rs [hunk 2]
```diff
@@ -352,6 +352,8 @@ declare_features! (
     (active, c_variadic, "1.34.0", Some(44930), None),
     /// Allows the use of `#[cfg(overflow_checks)` to check if integer overflow behaviour.
     (active, cfg_overflow_checks, "1.71.0", Some(111466), None),
+    /// Provides the relocation model information as cfg entry
+    (active, cfg_relocation_model, "1.71.0", Some(114929), None),
```
Comment (klensy) [inline compiler/rustc_feature/src/active.rs:356]: Shoudn't be `CURRENT_RUSTC_VERSION`?

Comment (bjorn3) [reply compiler/rustc_feature/src/active.rs:356]: Right

Comment (lu-zero) [reply compiler/rustc_feature/src/active.rs]: Amended :) Thank you.

File: tests/ui/feature-gates/feature-gate-cfg-relocation-model.rs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#[cfg(relocation_model = "pic")] //~ ERROR
+fn main() {}
```
Comment (bjorn3) [inline tests/ui/feature-gates/feature-gate-cfg-relocation-model.rs:2]:
```suggestion
#[cfg(relocation_model = "pic")] //~ ERROR
fn foo() {}

fn main() {}
```
This should prevent the extra error about `main` missing.

**Comment by bjorn3** [comment]
Looks like you did something wrong when rebasing. There are a couple of unrelated commits.

**Comment by lu-zero** [comment]
Let see if manually rebasing fixes it.

**Comment by lu-zero** [comment]
Hopefully no other targets are pic-impossible.

**Pull request merged**